### PR TITLE
feat(new source): Add Prometheus Pushgateway source

### LINKED ISSUE #10304: Prometheus Push Gateway Source
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Current Vector Version
```
0.16.1
```

### Use-cases

-  Use vector to accept Prometheus Push Gateway metrics requests 

We'd like to set up a vector cluster to  accept push metrics from lambda functions and cron jobs and fan them out to several sinks.

### Attempted Solutions

We can implement this currently with a combination of the http source and a custom lua filter. This is non ideal due to the performance penalty that using lua filters incur.   

### Proposal

Create a new Prometheus Push Gateway source that will accept the prometheus push gateway format and translate it into vector's internal metric format. We would like to be able to pull out some http headers on this request as well and add that as tags to the metric.

### PR DESCRIPTION
**Pull request by Sinjo**
I've just finished a first draft of the new Prometheus Pushgateway source for #10304.

I've given it a quick test by hand, using the [Ruby client](https://github.com/prometheus/client_ruby) and it seems to work as I was hoping. I'd love to get some feedback so I can polish this up and get it ready to merge.

Please excuse the messy commit history and the lack of tests. If you're +1 on the broad shape of this I'll fix both of those and address any feedback you give.

Fixes #10304

### CODE REVIEW

**Comment by neuronull** [comment]
Hi @Sinjo , thanks for proposing this new source.

I am curious if you have considered Stephen's earlier question: https://github.com/vectordotdev/vector/pull/18143#issuecomment-1677333355

> Please excuse the messy commit history and the lack of tests.

For the commit history- we will squash so no big deal. But if you would avoid force pushing that would be very helpful.

Regarding tests, yes we are good with the overall shape of this. I think the main thing missing right now is unit testing the parsing logic.
It would also be really great to see integration test added.

You will also want to run the following:

`make check-clippy` ,`make check-fmt` , make sure no errors.

`make generate-component-docs` , and commit the cue file additions (this needs to be done anytime the config struct is modified as well).

**Comment by Sinjo** [comment]
@StephenWakely Just to double check expectations: I'm happy to respond to issues, fix bugs that come up, and improve the docs if there are gaps (I'll write a first set of those as part of this PR). It'll be best effort, so I can't promise super fast timelines all the time, but I am happy to do it.

@neuronull I'll get to work on the rest of this PR soon and will avoid force pushing when I do. I've got a mixture of vacation and conference travel coming up so I'll be fitting the work around that.

**Comment by neuronull** [comment]
> @StephenWakely Just to double check expectations: I'm happy to respond to issues, fix bugs that come up, and improve the docs if there are gaps (I'll write a first set of those as part of this PR). It'll be best effort, so I can't promise super fast timelines all the time, but I am happy to do it.
> 
> @neuronull I'll get to work on the rest of this PR soon and will avoid force pushing when I do. I've got a mixture of vacation and conference travel coming up so I'll be fitting the work around that.

Thanks for letting us know on both fronts. Sounds good. Will keep an eye on the PR but let us know when it's ready!

**Comment by Sinjo** [comment]
Okay, I think this is ready for review.

One thing I'd like to do is document a couple of caveats:
  - That aggregation only works for counters and histograms (because it doesn't make sense for gauges or summaries)
  - That only `POST` is supported, because it's not possible to recreate the `PUT` behaviour of the real Pushgateway inside Vector's event architecture.

I saw that the intro text on pages like [this](https://vector.dev/docs/reference/configuration/sources/prometheus_scrape/) is pulled from the `description` field in a Markdown file. Is it okay to just have a slightly long description field or is there a better way to get longer text into that part of the docs?

**Review by urseberry — APPROVED** [review]

**Comment by StephenWakely** [comment]
> @StephenWakely Hey, sorry for the direct ping. I just wanted to check in on this PR and make sure it's not being filtered out of a review list by the `meta: awaiting author` tag.
> 
> I just merged upstream into my fork and updated my additions to match up with the module refactor. I gave it a quick test locally and everything still seems to be working fine.

Ah, awesome. Thanks. I'll take a look shortly.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Nice! I've run some manual tests and it is looking good. Thanks. There's some minor suggestions to the code that I think would be good to implement.

Also we need some additional files created for the documentation:

- We need a file in `/website/cue/reference/components/sources/prometheus_pushgateway.cue`.

You can have a look at some of the other prometheus files in that folder to get an idea of what needs to go there. There are some limitations to the source currently, which I think should be documented in the `how_it_works` section. Namely we currently don't support the `Put` Http Method and we don't support protobuf format - only text format is available at present.

- A markdown file is needed in `/website/content/en/docs/reference/configuration/sources/`.


If you aren't already doing so, would it be possible to run through all the checks mentioned [here](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#running-other-checks) just to make sure CI is happy when we try to merge.

File: src/sources/prometheus/pushgateway.rs [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+            return Err(ErrorMessage::new(
+                http::StatusCode::BAD_REQUEST,
+                "Job must not have an empty value".to_owned(),
+            ));
+        }
+
+        return Ok((stripped_key.to_owned(), "".to_owned()));
+    }
+
+    let decoded_bytes = BASE64_URL_SAFE.decode(v).map_err(|_| {
```
[the remark this answers is not in the document]
Comment (Sinjo) [reply src/sources/prometheus/pushgateway.rs]: Good catch! Just pushed a fix and I'm going to write the documentation next.

**Comment by Sinjo** [comment]
@StephenWakely Those are all done now, and I've run all the additional checks you linked to on my local machine. Hopefully this PR is ready to go!

I made my best attempt at the CUE file with the source documentation in, but it's possible I misunderstood one or two of the fields in there. Let me know if there's anything that needs changing.

File: src/sources/prometheus/pushgateway.rs [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+
+    #[configurable(derived)]
+    #[serde(default)]
+    keepalive: KeepaliveConfig,
+
+    /// Whether to aggregate values across pushes.
+    ///
+    /// Only applies to counters and histograms as gauges and summaries can't be
+    /// meaningfully aggregated.
+    #[configurable(metadata(docs::examples = true))]
```
Comment (StephenWakely) [inline src/sources/prometheus/pushgateway.rs:71]: It looks like examples are not supported on boolean fields.

```suggestion
```

File: website/cue/reference/components/sources/prometheus_pushgateway.cue [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		acknowledgements: true
+		multiline: enabled: false
+		receive: {
+			from: {
+				service: services.prometheus
+
+				interface: socket: {
+					api: {
+						title: "Prometheus Pushgateway"
+						url:   urls.prometheus_pushgateway
```
Comment (StephenWakely) [inline website/cue/reference/components/sources/prometheus_pushgateway.cue:26]: This `urls.prometheus_pushgateway` is referring to an entry in https://github.com/vectordotdev/vector/blob/master/website/cue/reference/urls.cue. But currently there is no such entry for `prometheus_pushgateway`. Would you be able to add an entry to point to https://github.com/prometheus/pushgateway/blob/master/README.md.

I'll be honest and admit that I'm not sure if this information is actually used anywhere in our documentation, but I could be missing something!

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Nice! Thank you. There's just a couple of documentation fixes that need doing now.

**Comment by Sinjo** [comment]
All done! I also added a commit sorting the `prometheus_*` URLs as a few of them seemed to be in the wrong spot and one was duplicated. Happy to drop that commit if that ordering was deliberate.

**Comment by StephenWakely** [comment]
@Sinjo Can you run `cue fmt` on the new cue files to ensure they are formatted correctly?

**Comment by Sinjo** [comment]
Done. Sorry about that. I had to leave the coworking space I was in yesterday in a bit of a rush and forgot to come back to that failure.

**Comment by StephenWakely** [comment]
That `check-changelog` CI failure looks like something has gone wrong on our end and is nothing to do with this PR. Hopefully we'll get that sorted out very soon.

**Comment by StephenWakely** [comment]
Hi. So unfortunately (or fortunately depending on your perspective!) we have just implemented a new requirement for adding changes to a change log. Given this PR is a user facing change, we are going to need an entry added the changelog.

You can find details of how to write to the changelog [here](https://github.com/vectordotdev/vector/tree/master/changelog.d). There aren't really any examples that I can find, beyond what is mentioned in that document given this is the first PR that will have one.

Is this something you can add?

**Comment by Sinjo** [comment]
@StephenWakely Done! Hopefully that looks right. I tried to follow all the guidelines from that document, but let me know if I missed something.

**Review by StephenWakely — APPROVED** [review]
Awesome! Thanks for seeing this one through @Sinjo. Really appreciate it.

**Pull request merged**